modules/cats/src/main/scala/pureconfig/module/cats/instances/package.scala

```diff
@@ -38,19 +26,44 @@ package object instances {
           }
         }
     }
-  }
 
-  implicit val configWriterCatsInstance: Contravariant[ConfigWriter] = new Contravariant[ConfigWriter] {
-    def contramap[A, B](fa: ConfigWriter[A])(f: B => A): ConfigWriter[B] =
-      fa.contramap(f)
-  }
+  implicit val configWriterCatsInstance: ContravariantSemigroupal[ConfigWriter] =
+    new ContravariantSemigroupal[ConfigWriter] {
+      def contramap[A, B](fa: ConfigWriter[A])(f: B => A): ConfigWriter[B] =
+        fa.contramap(f)
+
+      def product[A, B](fa: ConfigWriter[A], fb: ConfigWriter[B]) =
+        ConfigWriter.fromFunction[(A, B)] {
+          case (a, b) =>
+            fb.to(b).withFallback(fa.to(a))
+        }
+    }
+
+  implicit val configConvertCatsInstance: InvariantSemigroupal[ConfigConvert] =
+    new InvariantSemigroupal[ConfigConvert] {
+      def imap[A, B](fa: ConfigConvert[A])(f: A => B)(g: B => A): ConfigConvert[B] =
+        fa.xmap(f, g)
 
-  implicit val configConvertCatsInstance: Invariant[ConfigConvert] = new Invariant[ConfigConvert] {
-    def imap[A, B](fa: ConfigConvert[A])(f: A => B)(g: B => A): ConfigConvert[B] =
-      fa.xmap(f, g)
-  }
+      def product[A, B](fa: ConfigConvert[A], fb: ConfigConvert[B]): ConfigConvert[(A, B)] = {
+        val reader = fa.zip(fb)
+        val writer = ConfigWriter.fromFunction[(A, B)] {
+          case (a, b) =>
+            fb.to(b).withFallback(fa.to(a))
+        }
+
+        ConfigConvert.fromReaderAndWriter(
+          Derivation.Successful(reader),
+          Derivation.Successful(writer))
+      }
+    }
 
   implicit val configValueEq: Eq[ConfigValue] = Eq.fromUniversalEquals
   implicit val configReaderFailureEq: Eq[ConfigReaderFailure] = Eq.fromUniversalEquals
   implicit val configReaderFailuresEq: Eq[ConfigReaderFailures] = Eq.fromUniversalEquals
+
+  implicit val configReaderFailuresSemigroup: Semigroup[ConfigReaderFailures] =
+    Semigroup.instance(_ ++ _)
+
+  implicit val configValueCatsSemigroup: Semigroup[ConfigValue] =
+    Semigroup.instance((a, b) => b.withFallback(a))
```

Comment:
The order of this (ando also how `ConfigWriter`s are combined in the `product` methods) got me thinking. While I think it's natural to think of a deep merge from left to right, with subsequent properties overriding the assignments of previous ones, we tend to do it the other way around in PureConfig. For example, in the `loadConfigFromFiles` method, values in later files are used to backstop missing values from previous files. I think this difference of behavior might introduce a source of confusion. What do you guys think?

Reply:
Good point, indeed... Even though the merge operation on config values is defined as `withFallback`, I prefer the natural, map-like semantics of `++` (and I suppose cats syntax extensions also treat it as such). `loadConfigFromFiles` doesn't seem important/idiomatic enough to set the standard in my opinion, but I understand the concern.

Reply:
I suppose going with map-like semantics is OK for this case.